base_name_search_improved/models/ir_model.py

```diff
@@ -0,0 +1,260 @@
+                domain = [(rec_name, operator, name.replace(" ", "%"))]
+                res = _extend_name_results(
+                    self, base_domain + domain, res, limit, name_get_uid
+                )
+            # Try unordered word search on each of the search fields
+            # we only perform this search if we have at least one
+            # separator character
+            # also, if have raise the limit we skeep this iteration
+            if " " in name and len(res) < limit:
+                domain = []
+                for word in name.split():
+                    word_domain = []
+                    for rec_name in all_names:
+                        word_domain = (
+                            word_domain and ["|"] + word_domain or word_domain
+                        ) + [(rec_name, operator, word)]
+                    domain = (domain and ["&"] + domain or domain) + word_domain
+                res = _extend_name_results(
+                    self, base_domain + domain, res, limit, name_get_uid
+                )
+
+        return res
+
+    return _name_search
+
+
+def patch_fields_view_get():
+    @api.model
+    def fields_view_get(
+        self, view_id=None, view_type="form", toolbar=False, submenu=False
+    ):
+        res = fields_view_get.origin(
+            self,
+            view_id=view_id,
+            view_type=view_type,
+            toolbar=toolbar,
+            submenu=submenu,
+        )
+        if view_type == "search" and _get_add_smart_search(self):
+            eview = etree.fromstring(res["arch"])
+            placeholders = eview.xpath("//search/field")
+            if placeholders:
+                placeholder = placeholders[0]
+            else:
+                placeholder = eview.xpath("//search")[0]
+            placeholder.addnext(etree.Element("field", {"name": "smart_search"}))
+            eview.remove(placeholder)
+            res["arch"] = etree.tostring(eview)
+            res["fields"].update(self.fields_get(["smart_search"]))
+        return res
+
+    return fields_view_get
+
+
+class Base(models.AbstractModel):
+
+    _inherit = "base"
+
+    # TODO perhaps better to create only the field when enabled on the model
+    smart_search = fields.Char(
+        compute="_compute_smart_search", search="_search_smart_search",
+    )
+
+    def _compute_smart_search(self):
+        self.smart_search = False
+
+    @api.model
+    def _search_smart_search(self, operator, value):
+        """
+        Por ahora este método no llama a
+        self._name_search(name, operator=operator) ya que este no es tan
+        performante si se llama a ilimitados registros que es lo que el
+        name search debe devolver. Por eso se reimplementa acá nuevamente.
+        Además name_search tiene una lógica por la cual trata de devolver
+        primero los que mejor coinciden, en este caso eso no es necesario
+        Igualmente seguro se puede mejorar y unificar bastante código
```

Comment:
```suggestion
        For now this method does not call
        self._name_search(name, operator=operator) as this is not as
        performant if you call unlimited records which is what the
        name search should return. That is why it is reimplemented here again.
        In addition name_search has a logic by which it tries to return
        first the best matches, in this case that is not necessary.
        However, it can be improved and unify a lot of code.
```

Reply:
(automatic translation using Deepl.com)